### ISSUE TITLE
PyRight doesn't narrow type based on `hasattr`

### ISSUE THREAD

**Issue by robertmaxton42** [microsoft/pyright#6717]
**Describe the bug**
```Python
def foo(h: list | float):
    if hasattr(h, '__getitem__') and h[0] == 0:
        pass
```

PyRight puts a red squiggle under `h` in `h[0]`, saying "`"__getitem__" method not defined on type "float`". While canonically determining whether an object is indexable in Python is difficult enough that the Python documentation itself says that the only way to be sure is to try it and find out, in this case PyRight requires one specific method to exist, so it should also be able to tell that I've guaranteed that the method exists earlier in the line. It would be extra-nice if it could ascribe `h` to a duck type whose only known feature is that it has `__getitem__`, but at the least in this context it should be able to infer that, between the options `list` and `float`, `h` must be a `list`. (Moving `h[0] == 0` into a new `if` doesn't help.)

**VS Code extension or command-line**
PyLance v2023.11.10 in VSCode on WSL

**Comment by erictraut** [comment]
This isn't a supported type guard pattern, so no type narrowing is performed in this case. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

Each new type guard pattern requires specialized logic and potentially slows down overall type analysis. We do occasionally add support for new type guard patterns but only if we have strong signal that it's a common pattern. I don't think that `hasattr` qualifies, but we could consider adding this in the future if there's sufficient feedback from users. 

In the meantime, the recommended approach is to use a "user-defined type guard" using the `TypeGuard` mechanism introduced in [PEP 647](https://peps.python.org/pep-0647/).

I'm going to close this feature request for now because we have no current plans to add support for this type guard pattern.

**Comment by robertmaxton42** [comment]
I mean, `hasattr` is the thing the Python docs say you should do in performance-sensitive code in the `typing` docs themselves: see https://docs.python.org/3/library/typing.html#typing.runtime_checkable, the last gray box. So, to some extent it's "the correct thing to be doing"?

But sure, I guess I can use a `TypeGuard` for now.

**Comment by erictraut** [comment]
For comparison, mypy (another popular type checker) also does not support `hasattr` as a type guard pattern. I don't think this is a common idiom in Python.

The common approach is to use `isinstance`.

```python
def foo(h: list | float):
    if isinstance(h, list) and h[0] == 0:
        pass
```

The documentation you quoted applies to runtime protocols. The types you used in your example (`list` and `float`) are not protocol classes.

**Comment by robertmaxton42** [comment]
Well, sure, but my actual use case would have liked to use `isinstance(h, Iterable)` there, for example, and resorted to `hasattr` instead. (I suppose I should've used `hasattr(h, "__iter__")` too, oops.)

**Comment by MetRonnie** [comment]
I get that there may be reasons not to implement this, but this example is rather irritating
```python
def _pytest_passed(request: pytest.FixtureRequest) -> bool:
    """Returns True if the test(s) a fixture was used in passed."""
    if hasattr(request.node, '_function_outcome'):
        return request.node._function_outcome.outcome in {'passed', 'skipped'}
#                           ~~~~~~~~~~~~~~~~~
# Cannot access member "_function_outcome" for type "Item"
#  Member "_function_outcome" is unknownPylancereportAttributeAccessIssue
```

**Comment by Avasam** [comment]
Similarly I was surprised  that `hasattr` doesn't at least narrow `None` out of a union in this example: https://github.com/pypa/setuptools/pull/4262/files#diff-e3d3d454fa3a072c9f46f8affa27513892fbc2d245e87f57a5927a7be851de05R1639

**Comment by carlosgmartin** [comment]
For reference, here's a list of requests for this feature:
- https://github.com/microsoft/pyright/issues/5860
- https://github.com/microsoft/pyright/issues/5964
- https://github.com/microsoft/pyright/issues/6717
- https://github.com/microsoft/pyright/issues/6897
- https://github.com/microsoft/pyright/issues/7363
- https://github.com/microsoft/pylance-release/issues/2237

**Comment by Andrej730** [comment]
I was also confused when I've learned that there's no type narrowing on `hasattr` but now I think it's not a Pyright issue but more of a general Python typing system issue (issue, not a bug). I'll try to share to my understanding below, hopefully it will give others meeting this issue a different perspective.

The thing is, variable types typically act *covariantly*, meaning if some function argument has a type of `A` it means not just class `A` but also all possible subclasses of `A`. 

Even in the case outlined in the first post here, it is theoretically possible that someone will make up a weird class that will inherit from `float`, will implement `__getitem__` in a weird way that's not compatible with a `list`, making `foo` statically unsafe. So it is statically unsafe to assume `h.__getitem__` after `hasattr(h, '__getitem')` will be `list.__getitem__`.

```python
def foo(h: list | float):
    if hasattr(h, '__getitem__') and h[0] == 0:
        pass

class WeirdType(float):
    def __getitem__(self, key: str) -> str: ...

foo(WeirdType())
```

It is the same reason why Pyright or other type checkers is able recognize only always-true cases for `isinstance` but never only always-false - as type system implies there's always a theoretical possibility for the subtyped classes.
```python
# pyright: strict


def test(always_int: int) -> ...:
    # Pyright is able to recognize "always-true".
    # Unnecessary isinstance call; "int" is always an instance of "int".
    if isinstance(always_int, int):
        pass
    # No reports, pyright doesn't recognize it as "always-false".s
    if isinstance(always_int, str):
        # always_int: <subclass of int and str>
        reveal_type(always_int)
```

There are some possible ideas for `hasattr` improvements.
1. In non-strict typing mode, after `hasattr(obj, xxx)`, if one of possible types in `obj` doesn't have `xxx` attribute, then assume for it `obj.xxx` is `Any`. See example below. I guess it would help with the reported errors and would help with gradual typing but with any introduction of `Any`, it makes code not type safe.
```python
# Return type would be float | Any
def foo(h: list[float] | float):
    if hasattr(h, "__getitem__"):
        return h[0]
    raise Exception
```
2. Allow type narrowing in case if type is `final` and therefore it's not possible for it to `__getitem__` implemented.
```python
# pyright: strict
from typing import final

@final
class final_float: ...

# Return type would be `float`.
def foo(h: list[float] | final_float):
    if hasattr(h, "__getitem__"):
        return h[0]
    raise Exception
```
